### PR TITLE
Documentation rewrite for Random

### PR DESCRIPTION
**Pull request by gjp**
I'd like to request a full review. Please pick it apart. I'm not sure if I bit off more than I could chew for my first submission; this includes method-level corrections, examples, explanations, and some odd corner cases on Kernel.rand.

### CODE REVIEW

File: random.c [hunk 1]
```diff
@@ -450,20 +450,34 @@ rand_init(struct MT *mt, VALUE vseed)
...
+* The generator may be initialized with either a system-generated or user-supplied
+* seed value.
+*
+* Basic usage:
+*
+*   Random.rand       # a Float which is >= 0 and < 1
+*   Random.rand(100)  # a Fixnum in the range 0..99
+*
+* Note that there is another interface through Kernel.rand which behaves somewhat
+* differently.
```
Comment (semmons99) [inline random.c]: I would change this to, "Note that Kernel.rand behaves somewhat differently. It takes..." That way people have an idea if they should use this or that.

Comment (gjp) [inline random.c]: Kernel.rand's behavior became surprising to me after I read the existing docs. I'm not sure how to address this. My naive suggestion would be not to use it, but...it's Kernel, and it's probably doing what it's doing for a reason. Perhaps I should give more examples here, and illustrate by way of explaining what the Kernel version is missing?

**Comment by gjp** [comment]
This requires further editing. Kernel.rand is now a slightly broken version of Random.rand.

```
ruby-1.9.3-rc1 :127 > rand(1.8)
 => 0 
ruby-1.9.3-rc1 :128 > rand(0..1.8)
 => 0.6741722060911771
```

Backward compatibility or just backward?

Kernel.srand and Random.srand are the same thing, and messing with this seed value affects anything else which uses the default PRNG internally, such as Array#sample...

```
ruby-1.9.3-rc1 :129 > (1..1000).to_a.sample(4)
 => [951, 732, 598, 156] 
ruby-1.9.3-rc1 :130 > srand(42)
 => 42 
ruby-1.9.3-rc1 :131 > (1..1000).to_a.sample(4)
 => [375, 951, 732, 598] 
ruby-1.9.3-rc1 :132 > (1..1000).to_a.sample(4)
 => [157, 156, 58, 867] 
ruby-1.9.3-rc1 :133 > Random.srand(42)
 => 42 
ruby-1.9.3-rc1 :134 > (1..1000).to_a.sample(4)
 => [375, 951, 732, 598] 
```

so what I really ought to do is explain what Random::DEFAULT is, and then explain use cases for instancing Random instead of using DEFAULT, if I can find any.

**Comment by semmons99** [comment]
I like your changes. I think this is ready to be submitted to ruby core. You'll just need to squash your commits into a single commit and then generate a patch. Let me know if you need help.

**Comment by gjp** [comment]
I'm going to close this and see if I can clean up the mess I've made with git...

**Comment by gjp** [comment]
On Wed, Oct 5, 2011 at 6:40 AM, Shane Emmons <
reply@reply.github.com>wrote:

> I like your changes. I think this is ready to be submitted to ruby core.
> You'll just need to squash your commits into a single commit and then
> generate a patch. Let me know if you need help.

I submitted the patch using one of your patch submissions as a template :)

Quick question on Git. At least I hope it's quick:

It is considered bad practice to rebase commits which you have already
pushed to a remote repo. Doing so may invoke the wrath of the git gods...
you have been warned.

errrrm. Too late. Those commits were already pushed so they could be
reviewed. After a local rebase in order to squash 3 commits into one patch,
I am here:

# On branch trunk

# Your branch and 'origin/trunk' have diverged,

# and have 1 and 3 different commit(s) each, respectively.

#
nothing to commit (working directory clean)

Pushing won't work and pulling creates a mess. How do I recover from this,
and/or avoid this situation in the future?

Thanks!

## 

Gregory Parkhurst

**Comment by semmons99** [comment]
git push -f

**Comment by gjp** [comment]
On Wed, Oct 5, 2011 at 4:38 PM, Shane Emmons <
reply@reply.github.com>wrote:

> git push -f

that's it?  THAT'S IT??

_closes several tabs full of horribly abstruse Git commands_

## 

Gregory Parkhurst

**Pull request closed without merging**